testing/common/data/record/record_test.go

```diff
@@ -0,0 +1,763 @@
+import (
+	"context"
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/stretchr/testify/assert"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"kantoku/common/data/record"
+	"kantoku/impl/common/data/record/dumb"
+	mongorec "kantoku/impl/common/data/record/mongo"
+	"log"
+	"math/rand"
+	"sort"
+	"testing"
+)
+
+func newMongoRecords(ctx context.Context) *mongorec.Storage {
+	// Set connection configurations
+	clientOptions := options.Client().ApplyURI("mongodb://localhost:27017")
+
+	// Connect to the MongoDB server
+	client, err := mongo.Connect(ctx, clientOptions)
+	if err != nil {
+		log.Fatal(err)
+	}
+	go func() {
+		<-ctx.Done()
+
+		if err = client.Disconnect(ctx); err != nil {
+			log.Println("failed to disconnect from mongodb:", err)
+		}
+	}()
+
+	// Access the "main" database
+	database := client.Database("testing")
+
+	// Access the "testing" collection
+	collection := database.Collection("records")
+
+	return mongorec.New(collection)
+}
+
+var Keys = []string{
+	"updated_at",
+	"id",
+	"name",
+	"surname",
+	"group",
+}
+
+var Groups = []string{
+	"a",
+	"b",
+	"c",
+	"d",
+	"e",
+}
+
+const MinimumTestCases = 5
+const MaximumTestCases = 15
+const TotalRecords = 100
+
+func TestRecord(t *testing.T) {
+	rand.Seed(42)
+	ctx := context.Background()
+	implementations := map[string]record.Storage{
+		"mongo": newMongoRecords(ctx),
+	}
+
+	for label, _impl := range implementations {
+		impl := _impl
+
+		t.Run(label, func(t *testing.T) {
+			t.Log("Dataset initialization...")
+			t.Log("Erasing all data...")
+			if err := impl.Erase(ctx); err != nil {
+				t.Fatal("failed to erase:", err)
+			}
+
+			t.Log("Generating random records...")
+			specimen := dumb.New()
+			for i := 0; i < TotalRecords; i++ {
+				rec := randomRecord()
+				_ = specimen.Insert(ctx, rec)
+				if err := impl.Insert(ctx, rec); err != nil {
+					t.Fatal("failed to insert:", err)
+				}
+			}
+
+			t.Log("Generation successfully finished!")
+
+			t.Run("Read operations", func(t *testing.T) {
+				t.Run("Not filtered", func(t *testing.T) {
+					RunReadOperationsTests(ctx, t, impl, specimen)
+				})
+
+				t.Run("Filtered", func(t *testing.T) {
+					t.Run("Single filter", func(t *testing.T) {
+						RunFilterTests(ctx, t, [][][]record.Entry{
+							{
+								{{"group", "a"}},
+							},
+						}, impl, specimen)
+					})
+
+					t.Run("Multiple filters (one term)", func(t *testing.T) {
+						RunFilterTests(ctx, t, [][][]record.Entry{
+							{
+								lo.Map(Groups, func(group string, _ int) record.E {
+									return record.E{"group", group}
+								}),
+							},
+						}, impl, specimen)
+					})
+
+					t.Run("Multiple filters (multiple terms)", func(t *testing.T) {
+
+						var groupFilters []record.E
+						var nameFilters []record.E
+
+						for i := 0; i < 5; i++ {
+							groupFilters = append(groupFilters,
+								record.E{"group", lo.Sample(Groups)})
+						}
+
+						for i := 0; i < 100; i++ {
```

Comment:
should it be TotalRecords?

Reply:
it's just a random number from my head...